src/tactic/norm_cast.lean

```diff
@@ -25,225 +25,314 @@ try_for 1000 (mk_instance e)
+  `(@has_one.one %%β %%h1) ← return y,
+  h2 ← to_expr ``(has_one %%α) >>= mk_instance',
+  new_y ← to_expr ``( @coe %%α %%β %%coe1 (@has_one.one %%α %%h2) ),
+  eq_y ← aux_squash y new_y,
+  let new_e := app (app op x) new_y,
+  pr ← mk_congr_arg (app op x) eq_y,
+  return ((), new_e, pr)
+)
 | _ := failed
-end
 
 /-
-This is the main heuristic used alongside the elim_cast and move_cast lemmas.
-An expression of the shape: op (↑(x : α) : γ) (↑(y : β) : γ)
-is rewritten as:            op (↑(↑(x : α) : β) : γ) (↑(y : β) : γ)
-when the squash_cast lemmas can prove that (↑(x : α) : γ) = (↑(↑(x : α) : β) : γ)
+assumption is used to discharge proofs in step 2
 -/
-private meta def heur (_ : unit) (e : expr) : tactic (unit × expr × expr) :=
-match e with
-| (app (expr.app op x) y) :=
-do
-    `(@coe %%α %%δ %%coe1 %%xx) ← return x,
-    `(@coe %%β %%γ %%coe2 %%yy) ← return y,
-    success_if_fail $ is_def_eq α β,
-    is_def_eq δ γ,
-
-    (do
-        coe3 ← mk_app `has_lift_t [α, β] >>= mk_instance',
-        new_x ← to_expr ``(@coe %%β %%δ %%coe2 (@coe %%α %%β %%coe3 %%xx)),
-        let new_e := app (app op new_x) y,
-        eq_x ← aux_simp x new_x,
-        pr ← mk_congr_arg op eq_x,
-        pr ← mk_congr_fun pr y,
-        return ((), new_e, pr)
-    ) <|> (do
-        coe3 ← mk_app `has_lift_t [β, α] >>= mk_instance',
-        new_y ← to_expr ``(@coe %%α %%δ %%coe1 (@coe %%β %%α %%coe3 %%yy)),
-        let new_e := app (app op x) new_y,
-        eq_y ← aux_simp y new_y,
-        pr ← mk_congr_arg (app op x) eq_y,
-        return ((), new_e, pr)
-    )
-| _ := failed
-end
+private meta def prove : tactic unit :=
+assumption
 
 /-
-simpa is used to discharge proofs
+TODO: norm_cast takes a list of expressions to use as lemmas for the discharger
 -/
-private meta def prove : tactic unit :=
-tactic.interactive.simpa none ff [] [] none
 
+/-
+This is an auxiliary function used in step 2.
+It tries to rewrite an expression using the elim_cast and move_cast lemmas.
+On failure, it calls the heuristic.
+-/
 private meta def post (s : simp_lemmas) (_ : unit) (e : expr) : tactic (unit × expr × expr) :=
-do
-    r ← mcond (is_prop e) (return `iff) (return `eq),
-    (new_e, pr) ← s.rewrite e prove r,
-    pr ← match r with
-    |`iff := mk_app `propext [pr]
-    | _   := return pr
-    end,
-    return ((), new_e, pr)
+( do
+  r ← mcond (is_prop e) (return `iff) (return `eq),
+  (new_e, pr) ← s.rewrite e prove r,
+  pr ← match r with
+  | `iff := mk_app `propext [pr]
+  | _    := return pr
+  end,
+  return ((), new_e, pr)
+) <|> heur () e
+
+/-
+The following auxiliary functions are used to handle numerals.
+-/
 
+-- prove ↑n = n where n is a numeral
+private meta def aux_num_prove_eq (a b : expr) : tactic expr :=
+do
+  s1 ← simp_lemmas.mk_default,
+  s2 ← squash_cast_attr.get_cache,
+  let s := simp_lemmas.join s1 s2,
+  h ← to_expr ``(%%a = %%b),
+  (_, pr) ← simplify s [] h,
+  to_expr ``(eq.mpr %%pr trivial)
```

Comment:
This seems dangerous. You're assuming `pr` is a proof of `(a = b) = true`, right? What if `h` doesn't simplify all the way, so `pr : (a = b) = t`? If `t` is defeq to `true`, this could start an expensive proof by reflexivity.

I think the cleaner way to do this, starting after line 274, is
``` lean
   (_, pr) ← solve_aux h (simp_target s >> done),
   return pr
```
(or `prod.snd <$> solve_aux h (simp_target s >> done)`

The `done` forces that `simp` proves the goal `a = b` fully, instead of just reducing it.

Reply:
That's right, and at some point there definitely was a line to check the type of pr and the tactic failed if the term wasn't simplified enough.
I think I removed it with some debug traces, but using solve_aux is a lot cleaner anyway.

Reply:
I tried to replace this code to use simp_target and solve_aux instead, but I've hard some weird issues with simp_target creating new goals. The associated meta-variables crash infer_type, and I've not been able to write a minimal example of this behaviour.
For now I reverted to my version, with a check on the type of pr to make sure it is of the shape `_ = true`.